### ISSUE TITLE
exabgp should reject malformed subnets

### ISSUE THREAD

**Issue by chantra** [Exa-Networks/exabgp#715]
##### ISSUE TYPE

 - Bug Report

##### OS

```
Ubuntu 16.04.3 LTS
Uname  : Linux f733765bb4ee 4.9.41-moby #1 SMP Fri Aug 18 01:58:38 UTC 2017 x86_64
```

##### VERSION

```
root@f733765bb4ee:/# INTERPRETER=python /exabgp/sbin/exabgp --version
ExaBGP : 4.0.2-1c737d99
Python : 2.7.12 (default, Nov 19 2016, 06:48:10)  [GCC 5.4.0 20160609]
Uname  : Linux f733765bb4ee 4.9.41-moby #1 SMP Fri Aug 18 01:58:38 UTC 2017 x86_64
Root   : /exabgp
```

##### ENVIRONMENT 
```
root@f733765bb4ee:/# INTERPRETER=python /exabgp/sbin/exabgp --di

[exabgp.daemon]
drop = false
user = 'root'

[exabgp.tcp]
bind = '[0.0.0.0]'

```

##### CONFIGURATION
```
template {
  neighbor all {
    router-id 172.18.0.3;
    local-as 65001;
    hold-time 20;
    capability {
      graceful-restart;
    }
  }
}

neighbor 172.18.0.2 {
 inherit all;
  local-address 172.18.0.3;
  family {
    ipv4 unicast;
    ipv6 unicast;
  }
  peer-as 65000;
}
```

##### SUMMARY

Exabgp accepts malformed subnets, for instance, 1.1.1.1/24 will be automatically announced like 1.1.1.0/24.
This is dangerous and pretty much a good way to shoot oneself in the foot. I believe exabgp should be more opinionated and  reject such malformed subnets.

##### STEPS TO REPRODUCE

```
>>> CLIENT
# INTERPRETER=python /exabgp/bin/exabgpcli announce route 1.1.1.1/24 next-hop 172.18.0.3

>>> EXABGP LOGS
17:36:40 | 48     | process       | command from process api-internal-cli-1067d120 : announce route 1.1.1.1/24 next-hop 172.18.0.3
17:36:40 | 48     | reactor       | async | api-internal-cli-1067d120 | announce route 1.1.1.1/24 next-hop 172.18.0.3
17:36:40 | 48     | configuration | . route            | '1.1.1.1/24' 'next-hop' '172.18.0.3'
17:36:45 | 48     | api           | route added to neighbor 172.18.0.2 local-ip 172.18.0.3 local-as 65001 peer-as 65000 router-id 172.18.0.3 family-allowed in-open : 1.1.1.1/24 next-hop 172.18.0.3
17:36:45 | 48     | process       | responding to api-internal-cli-1067d120 : done
17:36:45 | 48     | outgoing-1    | received TCP payload (  19) FFFF FFFF FFFF FFFF FFFF FFFF FFFF FFFF 0013 04
17:36:45 | 48     | outgoing-1    | << message of type KEEPALIVE
17:36:45 | 48     | ka-outgoing-1 | send-timer -2 second(s) left
17:36:45 | 48     | outgoing-1    | sending TCP payload (  19) FFFF FFFF FFFF FFFF FFFF FFFF FFFF FFFF 0013 04
17:36:45 | 48     | outgoing-1    | >> KEEPALIVE
17:36:45 | 48     | outgoing-1    | sending TCP payload (  47) FFFF FFFF FFFF FFFF FFFF FFFF FFFF FFFF 002F 0200 0000 1440 0101 0040 0206 0201 0000 FDE9 4003 04AC 1200 0318 0101 01
17:36:45 | 48     | outgoing-1    | >> 1 UPDATE(s)
17:36:46 | 48     | ka-outgoing-1 | send-timer 5 second(s) left
17:36:46 | 48     | ka-outgoing-1 | receive-timer 18 second(s) left
17:36:47 | 48     | ka-outgoing-1 | send-timer 4 second(s) left

>>> CLIENT

# INTERPRETER=python /exabgp/bin/exabgpcli show adj-rib out
neighbor 172.18.0.2 ipv4 unicast 1.1.1.1/24 next-hop 172.18.0.3
>>> PEER
bird> show route
1.1.1.0/24         via 172.18.0.3 on eth0 [bgp1 17:36:45] * (100) [AS65001i]


>>> CLIENT
# INTERPRETER=python /exabgp/bin/exabgpcli announce route 1.1.1.2/24 next-hop 172.18.0.3

>>> EXABGP LOGS
.... no UPDATE sent as 1.1.1.0/24 is in the cache
17:38:02 | 48     | process       | command from process api-internal-cli-1067d120 : announce route 1.1.1.2/24 next-hop 172.18.0.3
17:38:02 | 48     | reactor       | async | api-internal-cli-1067d120 | announce route 1.1.1.2/24 next-hop 172.18.0.3
17:38:02 | 48     | configuration | . route            | '1.1.1.2/24' 'next-hop' '172.18.0.3'
17:38:05 | 48     | api           | route added to neighbor 172.18.0.2 local-ip 172.18.0.3 local-as 65001 peer-as 65000 router-id 172.18.0.3 family-allowed in-open : 1.1.1.2/24 next-hop 172.18.0.3
17:38:05 | 48     | process       | responding to api-internal-cli-1067d120 : done
17:38:05 | 48     | outgoing-1    | received TCP payload (  19) FFFF FFFF FFFF FFFF FFFF FFFF FFFF FFFF 0013 04
17:38:05 | 48     | outgoing-1    | << message of type KEEPALIVE
17:38:05 | 48     | ka-outgoing-1 | receive-timer 19 second(s) left
17:38:05 | 48     | ka-outgoing-1 | send-timer -2 second(s) left
17:38:05 | 48     | outgoing-1    | sending TCP payload (  19) FFFF FFFF FFFF FFFF FFFF FFFF FFFF FFFF 0013 04
17:38:05 | 48     | outgoing-1    | >> KEEPALIVE

>>> CLIENT
# INTERPRETER=python /exabgp/bin/exabgpcli show adj-rib out
neighbor 172.18.0.2 ipv4 unicast 1.1.1.1/24 next-hop 172.18.0.3

>>> PEER
bird> show route
1.1.1.0/24         via 172.18.0.3 on eth0 [bgp1 17:36:45] * (100) [AS65001i]



>>>> CLIENT
INTERPRETER=python /exabgp/bin/exabgpcli withdraw route 1.1.1.2/24 next-hop 172.18.0.3

>>>> EXABGP LOGS
17:42:56 | 48     | process       | command from process api-internal-cli-1067d120 : withdraw route 1.1.1.2/24 next-hop 172.18.0.3
17:42:56 | 48     | reactor       | async | api-internal-cli-1067d120 | withdraw route 1.1.1.2/24 next-hop 172.18.0.3
17:42:56 | 48     | configuration | . route            | '1.1.1.2/24' 'next-hop' '172.18.0.3'
17:42:56 | 48     | api           | route removed from neighbor 172.18.0.2 local-ip 172.18.0.3 local-as 65001 peer-as 65000 router-id 172.18.0.3 family-allowed in-open : 1.1.1.2/24 next-hop 172.18.0.3
17:42:56 | 48     | process       | responding to api-internal-cli-1067d120 : done
17:42:56 | 48     | outgoing-1    | sending TCP payload (  47) FFFF FFFF FFFF FFFF FFFF FFFF FFFF FFFF 002F 0200 0418 0101 0100 1440 0101 0040 0206 0201 0000 FDE9 4003 04AC 1200 03
17:42:56 | 48     | outgoing-1    | >> 1 UPDATE(s)
17:42:57 | 48     | ka-outgoing-1 | send-timer 2 second(s) left
17:42:57 | 48     | ka-outgoing-1 | receive-timer 17 second(s) left

>>>> PEER
... show route has no route


>>>> CLIENT
# INTERPRETER=python /exabgp/bin/exabgpcli show adj-rib out
has no route
```

##### EXPECTED RESULTS
I would expect that exabgp rejects malformed subnets rather than silently accepting and converting 1.1.1.1/24 into 1.1.1.0/24

##### ACTUAL RESULTS

1.1.1.1/24 get silently converted in 1.1.1.0/24 meaning that a user may unexpectedly advertise a different network than intended.

##### IMPORTANCE

This is a pretty bad behaviour as it silently ignore end user mistake and take a decision on the behalf of the user which may be wrong. Failing would be a better and more predictable behaviour.

For instance, if someone makes a typo and advertise ipv6 subnets:

`a:b:c:d1::/48` and `a:b:c:d2::/48` , first, exabgp will advertise `a:b:c::/48`, when the intent may have been to advertise `a:b:c:d1::/64` and `a:b:c:d2::/64`, we would end up advertising a bigger subnet.
Second, if user decides to withdraw `a:b:c:d2::/64`, the whole `a:b:c::/48` get withdrawn, which again may not be what is intended.
```

**Comment by thomas-mangin** [comment]
Hi @chantra 

`Be conservative in what you send, be liberal in what you accept` https://en.wikipedia.org/wiki/Robustness_principle ...

It is not a bug but I will consider it.

**Comment by chantra** [comment]
:) yeah, there is also

[Rule of Repair](https://en.wikipedia.org/wiki/Unix_philosophy#Eric_Raymond.E2.80.99s_17_Unix_Rules)
```
Developers should design programs that fail in a manner that is easy to localize
and diagnose or in other words “fail noisily”. This rule aims to prevent incorrect
output from a program from becoming an input and corrupting the output of other
code undetected.
```

Actually, even in the robustness principle, `code that receives input should accept non-conformant input as long as the meaning is clear.`. In this very case, it is not clear if the user intend to advertise a /32 and made a mistake in the netmask, or intended to advertise a /24 but made a mistake in the network prefix.

Basically, there is no way to make the right decision here. Failing early (as in not advertising and logging that the input in incorrect/ambiguous) would allow people to fix the issue rather than having things apparently working as a side-effect.

I started going down the rabbit hole of isolating where this happen, but this part of the code seems pretty complex to follow. In any cases happy to dig more into it if you think it is a good idea.

**Comment by thomas-mangin** [comment]
Yes, I call that principle `fail early, fail loud`. I will think on how it could be implemented. It is still possible for 4.0 as no stable release is out yet.

I will not change the behaviour for 3.4 (it may cause regression).

**Comment by chantra** [comment]
Yes, I call that principle fail early, fail loud. I will think on how it
could be implemented. It is still possible for 4.0 as no stable release is
out yet.

Thanks. I suppose the validation is fairly simple, one of the main issue is
that afaict, there is no communication of success/failure between exabgp
and the processes sending the commands. That being said, i think a first
step of validating and logging the error would be a good enough approach in
making the user aware of their mistakes.

I will not change the behaviour for 3.4 (it may cause regression).

Sounds reasonable to me.

**Comment by thomas-mangin** [comment]
4.0 does ack command with an "ok" of "failed" message (which can be turned off to reduce the burden of porting from 3.4)

**Comment by thomas-mangin** [comment]
better late than never ..
